libcudacxx/include/cuda/__barrier/barrier_block_scope.h

```diff
@@ -87,140 +94,184 @@ class barrier<thread_scope_block, ::cuda::std::__empty_completion> : public __bl
-          ::__trap();
-        }),
-      NV_PROVIDES_SM_80,
-      (if (::cuda::device::is_object_from(__barrier, ::cuda::device::address_space::shared)) {
-        asm volatile("mbarrier.inval.shared.b64 [%0];" ::"r"(static_cast<::cuda::std::uint32_t>(
-          ::__cvta_generic_to_shared(&__barrier)))
-                     : "memory");
-      }))
+      (if (::cuda::device::is_object_from(__barrier, ::cuda::device::address_space::cluster_shared)) { ::__trap(); }))
   }
 
   _CCCL_API inline friend void init(barrier* __b,
                                     ::cuda::std::ptrdiff_t __expected,
                                     ::cuda::std::__empty_completion = ::cuda::std::__empty_completion())
   {
-    NV_DISPATCH_TARGET(
-      NV_PROVIDES_SM_90,
-      (
-        if (::cuda::device::is_object_from(__b->__barrier, ::cuda::device::address_space::shared)) {
-          asm volatile("mbarrier.init.shared.b64 [%0], %1;" ::"r"(
-                         static_cast<::cuda::std::uint32_t>(::__cvta_generic_to_shared(&__b->__barrier))),
-                       "r"(static_cast<::cuda::std::uint32_t>(__expected))
-                       : "memory");
-        } else if (::cuda::device::is_object_from(__b->__barrier, ::cuda::device::address_space::cluster_shared)) {
-          ::__trap();
-        } else { new (&__b->__barrier) __barrier_base(__expected); }),
-      NV_PROVIDES_SM_80,
-      (
-        if (::cuda::device::is_object_from(__b->__barrier, ::cuda::device::address_space::shared)) {
-          asm volatile("mbarrier.init.shared.b64 [%0], %1;" ::"r"(
-                         static_cast<::cuda::std::uint32_t>(::__cvta_generic_to_shared(&__b->__barrier))),
-                       "r"(static_cast<::cuda::std::uint32_t>(__expected))
-                       : "memory");
-        } else { new (&__b->__barrier) __barrier_base(__expected); }),
-      NV_ANY_TARGET,
-      (new (&__b->__barrier) __barrier_base(__expected);))
+    NV_IF_TARGET(NV_PROVIDES_SM_80,
+                 (if (::cuda::device::is_object_from(__b->__barrier, ::cuda::device::address_space::shared)) {
+                   ::cuda::ptx::mbarrier_init(__b->__native_handle(), static_cast<::cuda::std::uint32_t>(__expected));
+                   return;
+                 }))
+
+    NV_IF_TARGET(NV_PROVIDES_SM_90,
+                 (if (::cuda::device::is_object_from(__b->__barrier, ::cuda::device::address_space::cluster_shared)) {
+                   ::__trap();
+                 }))
+
+    new (&__b->__barrier) __barrier_base(__expected);
+  }
+
+private:
+  _CCCL_DEVICE_API _CCCL_FORCEINLINE arrival_token __arrive_sm90(::cuda::std::ptrdiff_t __update)
+  {
+    if (!::cuda::device::is_object_from(__barrier, ::cuda::device::address_space::cluster_shared))
+    {
+      return __barrier.arrive(__update);
+    }
+    if (!::cuda::device::is_object_from(__barrier, ::cuda::device::address_space::shared))
+    {
+      ::__trap();
+    }
```

Comment:
This is strange, because the first condition takes anything but `cluster_shared`, so the second one seems wrong

Reply:
Anything that is `shared` is also `cluster_shared`. Because the shared memory address space is part of the cluster shared memory space.

Reply:
Replaced by `_CCCL_ASSERT`